samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -0,0 +1,319 @@
+  // the condition.
+  private void getStreamAndTableKeyIds(List<RexNode> operands, final LogicalJoin join, boolean isTablePosOnRight,
+      List<Integer> streamIds, List<Integer> tableIds) {
+
+    // All non-leaf operands in the join condition should be expressions.
+    if (operands.get(0) instanceof RexCall) {
+      operands.forEach(operand -> {
+        validateJoinCondition(operand);
+        getStreamAndTableKeyIds(((RexCall) operand).getOperands(), join, isTablePosOnRight, streamIds, tableIds);
+      });
+      return;
+    }
+
+    // We are at the leaf of the join condition. Only binary operators are supported.
+    Validate.isTrue(operands.size() == 2);
+
+    // Only reference operands are supported in row expressions and not constants.
+    // a.key = b.key is supported with a.key and b.key being reference operands.
+    // a.key = "constant" is not yet supported.
+    if (!(operands.get(0) instanceof RexInputRef) || !(operands.get(1) instanceof RexInputRef)) {
+      throw new SamzaException("SQL query is not supported. Join condition " + join.getCondition() + " should have "
+          + "reference operands but the types are " + operands.get(0).getClass() + " and " + operands.get(1).getClass());
+    }
+
+    // Join condition is commutative, meaning, a.key = b.key is equivalent to b.key = a.key.
+    // Calcite assigns the indices to the fields based on the order a and b are specified in
+    // the sql 'from' clause. Let's put the operand with smaller index in leftRef and larger
+    // index in rightRef.
+    RexInputRef leftRef = (RexInputRef) operands.get(0);
+    RexInputRef rightRef = (RexInputRef) operands.get(1);
+    if (leftRef.getIndex() > rightRef.getIndex()) {
+      RexInputRef tmpRef = leftRef;
+      leftRef = rightRef;
+      rightRef = tmpRef;
+    }
+
+    // Get the table index and stream index
+    int deltaIdx = rightRef.getIndex() - join.getLeft().getRowType().getFieldCount();
+    streamIds.add(isTablePosOnRight ? leftRef.getIndex() : deltaIdx);
+    tableIds.add(isTablePosOnRight ? deltaIdx : leftRef.getIndex());
+  }
+
+  private String dumpRelPlanForNode(RelNode relNode) {
+    return RelOptUtil.dumpPlan("Rel expression: ",
+        relNode, SqlExplainFormat.TEXT,
+        SqlExplainLevel.EXPPLAN_ATTRIBUTES);
+  }
+
+  private boolean isTable(RelNode relNode) {
+    // NOTE: Any intermediate form of a join is always a stream. Eg: For the second level join of
+    // stream-table-table join, the left side of the join is join output, which we always
+    // assume to be a stream. The intermediate stream won't be an instance of EnumerableTableScan.
+    return relNode instanceof EnumerableTableScan &&
+        sourceResolver.isTable(String.join(".", relNode.getTable().getQualifiedName()));
+  }
+
+  private final class SamzaSqlRelMessageJoinFunction implements StreamTableJoinFunction<SamzaSqlCompositeKey,
```

Comment:
Consider separating this into a separate class, So that we can add unit tests targeting specifically the logic in this class.